sw/airborne/modules/calibration/mag_calib_ukf.c

```diff
@@ -0,0 +1,199 @@
+  if (update_geo_mag_field && geo_mag.ready) {
+    double n                = double_vect3_norm(&geo_mag.vect);
+    if (n > 0.01) {
+      H.x                     = (float) geo_mag.vect.x / n;
+      H.y                     = (float) geo_mag.vect.y / n;
+      H.z                     = (float) geo_mag.vect.z / n;
+      mag_field_from_geo_mag  = time(0);
+      update_geo_mag_field    = false;
+      VERBOSE_PRINT("Updating local magnetic field from geo_mag module (Hx: %4.2f, Hy: %4.2f, Hz: %4.2f)\n", H.x, H.y, H.z);
+    }
+  }
+  if (MAG_CALIB_UKF_GEO_MAG_TIMEOUT && GpsFixValid() && difftime(time(0), mag_field_from_geo_mag) >= MAG_CALIB_UKF_GEO_MAG_TIMEOUT) {
+    geo_mag.ready           = false;
+    geo_mag.calc_once       = true;   ///< Geo_mag will not re-update the calculation when the throttle is on so this is neccesary
+    update_geo_mag_field    = true;
+  }
+  /** Update magnetometer UKF and calibrate measurement **/
+  if (mag->x != 0 || mag->y != 0 || mag->z != 0) {
+    /** Rotate the local magnetic field by our current attitude **/
+    struct FloatQuat *body_quat = stateGetNedToBodyQuat_f();
+    struct FloatRMat body_rmat;
+    float_rmat_of_quat(&body_rmat, body_quat);
+    struct FloatVect3 expected_measurement;
+    float_rmat_vmult(&expected_measurement, &body_rmat, &H);
+    float expected_mag_field[3] = { expected_measurement.x, expected_measurement.y, expected_measurement.z };
+    /** Update magnetometer UKF **/
+    measurement[0] = MAG_FLOAT_OF_BFP(mag->x);
+    measurement[1] = MAG_FLOAT_OF_BFP(mag->y);
+    measurement[2] = MAG_FLOAT_OF_BFP(mag->z);
+    TRICAL_estimate_update(&mag_calib, measurement, expected_mag_field);
+    TRICAL_measurement_calibrate(&mag_calib, measurement, calibrated_measurement);
+    /** Save calibrated result **/
+    mag->x = (int32_t) MAG_BFP_OF_REAL(calibrated_measurement[0]);
+    mag->y = (int32_t) MAG_BFP_OF_REAL(calibrated_measurement[1]);
+    mag->z = (int32_t) MAG_BFP_OF_REAL(calibrated_measurement[2]);
+    VERBOSE_PRINT("magnetometer measurement (x: %4.2f  y: %4.2f  z: %4.2f) norm: %4.2f\n", measurement[0], measurement[1], measurement[2], hypot(hypot(measurement[0], measurement[1]), measurement[2]));
+    VERBOSE_PRINT("magnetometer bias_f      (x: %4.2f  y: %4.2f  z: %4.2f)\n", mag_calib.state[0], mag_calib.state[1],  mag_calib.state[2]);
+    VERBOSE_PRINT("expected measurement     (x: %4.2f  y: %4.2f  z: %4.2f) norm: %4.2f\n", expected_mag_field[0], expected_mag_field[1], expected_mag_field[2], hypot(hypot(expected_mag_field[0], expected_mag_field[1]), expected_mag_field[2]));
+    VERBOSE_PRINT("calibrated   measurement (x: %4.2f  y: %4.2f  z: %4.2f) norm: %4.2f\n\n", calibrated_measurement[0], calibrated_measurement[1], calibrated_measurement[2], hypot(hypot(calibrated_measurement[0], calibrated_measurement[1]), calibrated_measurement[2]));
+  }
+  AbiSendMsgIMU_MAG_INT32(MAG_CALIB_UKF_ID, get_sys_time_usec(), mag);
```

Comment:
it would be better to resend the message with the original time stamp instead of the time when the computation is done